### PR TITLE
SNOW-662338: Handle `None` returned by connector's cursor

### LINKED ISSUE #496: Long running program iterating over data with to_local_iterator crashing unexpectedly
Please answer these questions before submitting your issue. Thanks!

1. What version of Python are you using?

```
Python 3.8.9 (default, Apr  6 2022, 10:43:36)
[GCC 9.4.0]
```

2. What operating system and processor architecture are you using?

```
Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.29
```

3. What are the component versions in the environment (`pip freeze`)?

```
snowflake-snowpark-python==0.9.0
```

4. What did you do?

To summarize, I have a long running program iterating over a huge table. This job performs a lot of processing on the data for every iteration. The code looks something like this:

```python
import gc

from snowflake.snowpark import (
    DataFrame,
    Session
)

from snowflake.snowpark.row import Row

snowflake_connect_options = {
   "account": os.environ['SNOWFLAKE_ACCOUNT'],
   "user": os.environ['SNOWFLAKE_USER'],
   "password": os.environ['SNOWFLAKE_PASSWORD'],
   "warehouse": os.environ['SNOWFLAKE_WAREHOUSE'],
   "role": os.environ['SNOWFLAKE_ROLE'],
   "database": os.environ['SNOWFLAKE_DATABASE'],
   "client_session_keep_alive": True,
}

session = Session.builder.configs(snowflake_connect_options).create()

local_iterator = self.session.table(qualified_raw_table).to_local_iterator()

for row_data in local_iterator:
   # Perform processing operations on row_data
   gc.collect()
```

5. What did you expect to see?

I expect the code to run through the local_iterator without issue, instead this happened after 6 hours (meaning it's not the 4 hour timeout messing with me):

```
File "/path/to/code.py", line X, in code_function
   for row_data in local_iterator:
File "/usr/local/lib/python3.8/site-packages/snowflake/snowpark/dataframe.py", line 505, in to_local_iterator
   yield from self._session._conn.execute(
File "/usr/local/lib/python3.8/site-packages/snowflake/snowpark/_internal/utils.py", line 478, in result_set_to_iter
   row = Row(*data)
TypeError: type object argument after * must be an iterable, not NoneType
```

6. Can you set logging to DEBUG and collect the logs?

Unfortunately it's hard for me to do this as the job takes a long time to run and generates a lot of other logs.

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

**Review by sfc-gh-aalam — COMMENTED** [review]
The change looks good. It is possible to add a test case that checks a None return and a non-None return

**Comment by sfc-gh-jdu** [comment]
> The change looks good. It is possible to add a test case that checks a None return and a non-None return

Yes, I wish to, but unfortunately I'm not able to reproduce the case returning None. Hope @sfc-gh-mkeller will know more about it.

**Comment by sfc-gh-jdu** [comment]
> > > The change looks good. It is possible to add a test case that checks a None return and a non-None return
> > 
> > 
> > Yes, I wish to, but unfortunately I'm not able to reproduce the case returning None. Hope @sfc-gh-mkeller will know more about it.
> 
> How about a unit test using `mock`?

good point, I'll try to add

**Review by sfc-gh-sfan — APPROVED** [review]

**Review by sfc-gh-aalam — APPROVED** [review]

File: src/snowflake/snowpark/_internal/utils.py [hunk 1]
```diff
@@ -519,12 +519,12 @@ def result_set_to_rows(
         col_names = [col.name for col in result_meta]
         rows = []
         for data in result_set:
-            row = Row(*data)
+            row = Row() if data is None else Row(*data)
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/_internal/utils.py:522]: I'm thinking we should gracefully alert users when data is `None` so they know the reason of the error and how to avoid it.

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/utils.py:522]: logger.warning? However, we don't know the exact reason why it can be None (as we chatted yesterday, there might be multiple reasons).

Comment (sfc-gh-yixie) [reply src/snowflake/snowpark/_internal/utils.py:522]: We may need to update python connector. Does connector have more information to tell the root cause?

Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/_internal/utils.py:522]: not yet, see here https://github.com/snowflakedb/snowpark-python/issues/496#issuecomment-1284470345

**Review by sfc-gh-yixie — APPROVED** [review]
